pennylane/devices/tests/test_gates_with_expval.py

```diff
@@ -0,0 +1,364 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Tests that application of gates and state preparations works correctly on a
+device by checking expectation values.
+"""
+# pylint: disable=no-self-use
+# pylint: disable=too-many-arguments
+# pylint: disable=pointless-statement
+from math import sqrt, pi
+
+import pytest
+import numpy as np
+import pennylane as qml
+
+from flaky import flaky
+
+pytestmark = pytest.mark.skip_unsupported
+
+np.random.seed(42)
+
+
+# ===============================================================
+
+
+@flaky(max_runs=10)
+class TestGatesQubitExpval:
+    """Test some expectation values obtained from qubit-based devices after
+    application of gates."""
+
+    # This test is ran with two Z expvals
+    @pytest.mark.parametrize(
+        "par,wires,expected_output",
+        [([1, 1], [0, 1], [-1, -1]), ([1], [0], [-1, 1]), ([1], [1], [1, -1])],
+    )
+    def test_basis_state_2_qubit_subset(self, device, tol, par, wires, expected_output):
+        """Tests qubit basis state preparation on subsets of qubits"""
+        n_wires = 2
+        dev = device(n_wires)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.BasisState(np.array(par), wires=wires)
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        assert np.allclose(circuit(), expected_output, atol=tol(dev.shots))
+
+    # This test is run with three expvals
+    @pytest.mark.parametrize(
+        "par,wires,expected_output",
+        [
+            (
+                [1j / np.sqrt(10), (1 - 2j) / np.sqrt(10), 0, 0, 0, 2 / np.sqrt(10), 0, 0],
+                [0, 1, 2],
+                [1 / 5.0, 1.0, -4 / 5.0],
+            ),
+            ([1 / np.sqrt(2), 0, 0, 1 / np.sqrt(2)], [0, 2], [0.0, 1.0, 0.0]),
+            ([1 / np.sqrt(2), 0, 0, 1 / np.sqrt(2)], [0, 1], [0.0, 0.0, 1.0]),
+            ([0, 1, 0, 0, 0, 0, 0, 0], [2, 1, 0], [-1.0, 1.0, 1.0]),
+            ([0, 1j, 0, 0, 0, 0, 0, 0], [0, 2, 1], [1.0, -1.0, 1.0]),
+            ([0, 1 / np.sqrt(2), 0, 1 / np.sqrt(2)], [1, 0], [-1.0, 0.0, 1.0]),
+            ([0, 1 / np.sqrt(2), 0, 1 / np.sqrt(2)], [0, 1], [0.0, -1.0, 1.0]),
```

Comment:
clever test cases!

Reply:
courtesy of the `default.qubit` test suite :slightly_smiling_face: